Add SettableGauge interface and an implementation

xiaochuanyu opened the pull request:
This PR attempts to address https://github.com/dropwizard/metrics/issues/1126.

Example usage:
```
MetricRegistry registry = new MetricRegistry();
SettableGauge<Integer> settable = registry.register("example1", new SimpleSettableGauge<>(0));
settable.setValue(123);
```
Some things I want to point out that I'm not sure about:
1. Whats a better name than `SimpleSettableGauge`?
2. I'm not sure if we need a `MetricRegistry.settableGauge`.

dependabot-preview[bot] committed:
Create Dependabot config file (#1592)

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump maven-project-info-reports-plugin from 3.0.0 to 3.1.0 (#1598)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump maven-shade-plugin from 3.2.3 to 3.2.4 (#1597)

Bumps [maven-shade-plugin](https://github.com/apache/maven-shade-plugin) from 3.2.3 to 3.2.4.
- [Release notes](https://github.com/apache/maven-shade-plugin/releases)
- [Commits](https://github.com/apache/maven-shade-plugin/compare/maven-shade-plugin-3.2.3...maven-shade-plugin-3.2.4)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

xiaochuanyu committed:
Add SettableGauge interface and an implementation

the-thing commented:
Just a suggestion. Changing the metrics registry signature seems to be backwards compatible, but it will require some unit tests.

`public <T extends Gauge> T gauge(String name, final MetricSupplier<T> supplier);`

the-thing committed:
ExponentiallyDecayingReservoirTest improvements (#1587)

Improves some of the comments and test parameters in `com.codahale.metrics.ExponentiallyDecayingReservoirTest`.

I think that https://github.com/dropwizard/metrics/issues/1582 can be closed after this is merged as the implementation is correct and statistical error is expected.

The increase of alpha value for `longPeriodsOfInactivityShouldNotCorruptSamplingState` should eliminate/reduce failure. I couldn't recrate this value.

xiaochuanyu committed:
Update MetricRegistry.gauge() signature to be more generic

xiaochuanyu commented:
@the-thing I updated `guage()`'s signature per your suggestion 99c888f.

joschi commented:
@xiaochuanyu Please change the base branch to `release/4.2.x` and rebase the PR if necessary.

the-thing commented on metrics-core/src/main/java/com/codahale/metrics/SimpleSettableGauge.java:
```diff
@@ -0,0 +1,44 @@
+public class SimpleSettableGauge<T> implements SettableGauge<T> {
+    /**
+     * Current value.
+     * Volatile so that assignment is thread-safe.
```
You don't need to explain why volatile is there (people should know it). Probably the whole field comment is unnecessary.

the-thing commented on metrics-core/src/main/java/com/codahale/metrics/MetricRegistry.java:
```diff
@@ -310,10 +310,10 @@ public boolean isInstance(Metric metric) {
     @SuppressWarnings("rawtypes")
-    public Gauge gauge(String name, final MetricSupplier<Gauge> supplier) {
-        return getOrAdd(name, new MetricBuilder<Gauge>() {
+    public <T extends  Gauge> T gauge(String name, final MetricSupplier<T> supplier) {
```
Formatting (extra space).

the-thing commented on metrics-core/src/main/java/com/codahale/metrics/SimpleSettableGauge.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * <p>
+ * Similar to {@link Gauge}, but metric value is updated via calling {@link #setValue(T)} instead.
+ * Thread-safety is ensured via storing metric value as a volatile member variable.
```
The thread safety aspect doesn't need to be explained.

xiaochuanyu commented:
Hmm, I can't edit the source branch without opening a separate PR which is current `xiaochuanyu:release/4.1.x` but I want `xiaochuanyu:settable-gauge`.
I think I'm just gonna make another PR and close this one.
